src/JsonSchema/SchemaFactory.php

```diff
@@ -192,8 +205,26 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
             $propertySchema['example'] = $propertySchema['default'];
         }
 
+        // never override the following keys if at least one is already set (by user or restrictions)
+        if (
+            0 < \count(
+                array_intersect(
+                    ['type', '$ref', 'anyOf', 'allOf', 'oneOf'],
+                    array_keys($propertySchema)
+                )
+            )
```

Comment:
I wish we could do something a little bit more optimized to avoid 3 function calls and a `count`:

```
if ($propertySchema['type'] ?? $propertySchema['$ref'] ?? $propertySchema['anyOf'] ?? $propertySchema['allOf'] ?? $propertySchema['oneOf'] ?? false) {
```